Compilation error when building with --cfg doc

Is there some known limitation regarding generating the documentation of `socket2`?

I tried generating the documentation *with links to definitions* of a crate depending on `socket2`:

```
cargo init
cargo add socket2@0.4.7
RUSTDOCFLAGS="--generate-link-to-definition -Zunstable-options" cargo +nightly-2023-01-15 doc
```

I expected the `cargo doc` above to succeed. Instead, I get some compilation errors:
```text
   Compiling libc v0.2.139
 Documenting libc v0.2.139
    Checking socket2 v0.4.7
 Documenting socket2 v0.4.7
error[E0433]: failed to resolve: could not resolve path `sys::socketpair`
   --> /home/fpoli/.cargo/registry/src/github.com-1ecc6299db9ec823/socket2-0.4.7/src/socket.rs:172:9
    |
172 |         sys::socketpair(domain.0, ty.0, protocol)
    |         ^^^^^^^^^^^^^^^ could not resolve path `sys::socketpair`
    |
    = note: this error was originally ignored because you are running `rustdoc`
    = note: try running again with `rustc` or `cargo check` and you may get a more detailed error

error[E0433]: failed to resolve: could not resolve path `sys::keepalive_time`
    --> /home/fpoli/.cargo/registry/src/github.com-1ecc6299db9ec823/socket2-0.4.7/src/socket.rs:1663:9
     |
1663 |         sys::keepalive_time(self.as_raw())
     |         ^^^^^^^^^^^^^^^^^^^ could not resolve path `sys::keepalive_time`
     |
     = note: this error was originally ignored because you are running `rustdoc`
     = note: try running again with `rustc` or `cargo check` and you may get a more detailed error

For more information about this error, try `rustc --explain E0433`.
error: could not document `socket2`
```

Even without using nightly Rust, the following fails to compile (suggested in https://github.com/rust-lang/rust/issues/107063):
```
RUSTFLAGS="--cfg doc" cargo build
```
```
   Compiling libc v0.2.139
   Compiling socket2 v0.4.7
error[E0425]: cannot find function `socketpair` in module `sys`
   --> /home/fpoli/.cargo/registry/src/github.com-1ecc6299db9ec823/socket2-0.4.7/src/socket.rs:172:14
    |
172 |         sys::socketpair(domain.0, ty.0, protocol)
    |              ^^^^^^^^^^ not found in `sys`
    |
help: consider importing this function
    |
9   | use libc::socketpair;
    |
help: if you import `socketpair`, refer to it directly
    |
172 -         sys::socketpair(domain.0, ty.0, protocol)
172 +         socketpair(domain.0, ty.0, protocol)
    |

error[E0425]: cannot find function `keepalive_time` in module `sys`
    --> /home/fpoli/.cargo/registry/src/github.com-1ecc6299db9ec823/socket2-0.4.7/src/socket.rs:1663:14
     |
1663 |         sys::keepalive_time(self.as_raw())
     |              ^^^^^^^^^^^^^^ help: a constant with a similar name exists: `KEEPALIVE_TIME`
     |
    ::: /home/fpoli/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.139/src/unix/linux_like/mod.rs:964:1
     |
964  | pub const TCP_KEEPIDLE: ::c_int = 4;
     | ------------------------------- similarly named constant `KEEPALIVE_TIME` defined here

For more information about this error, try `rustc --explain E0425`.
error: could not compile `socket2` due to 2 previous errors
```

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -512,7 +512,6 @@ impl TcpKeepalive {
     #[cfg(all(
         feature = "all",
         any(
-            doc,
             target_os = "android",
             target_os = "dragonfly",
             target_os = "freebsd",

--- a/src/socket.rs
+++ b/src/socket.rs
@@ -147,7 +147,7 @@ impl Socket {
     /// This function sets the same flags as in done for [`Socket::new`],
     /// [`Socket::pair_raw`] can be used if you don't want to set those flags.
     #[doc = man_links!(unix: socketpair(2))]
-    #[cfg(any(doc, all(feature = "all", unix)))]
+    #[cfg(all(feature = "all", unix))]
     #[cfg_attr(docsrs, doc(cfg(all(feature = "all", unix))))]
     pub fn pair(
         domain: Domain,
@@ -164,7 +164,7 @@ impl Socket {
     /// Creates a pair of sockets which are connected to each other.
     ///
     /// This function corresponds to `socketpair(2)`.
-    #[cfg(any(doc, all(feature = "all", unix)))]
+    #[cfg(all(feature = "all", unix))]
     #[cfg_attr(docsrs, doc(cfg(all(feature = "all", unix))))]
     pub fn pair_raw(
         domain: Domain,
@@ -1091,7 +1091,7 @@ impl Socket {
     /// For more information about this option, see [`set_ip_transparent`].
     ///
     /// [`set_ip_transparent`]: Socket::set_ip_transparent
-    #[cfg(any(doc, all(feature = "all", target_os = "linux")))]
+    #[cfg(all(feature = "all", target_os = "linux"))]
     #[cfg_attr(docsrs, doc(cfg(all(feature = "all", target_os = "linux"))))]
     pub fn ip_transparent(&self) -> io::Result<bool> {
         unsafe {
@@ -1115,7 +1115,7 @@ impl Socket {
     ///
     /// TProxy redirection with the iptables TPROXY target also
     /// requires that this option be set on the redirected socket.
-    #[cfg(any(doc, all(feature = "all", target_os = "linux")))]
+    #[cfg(all(feature = "all", target_os = "linux"))]
     #[cfg_attr(docsrs, doc(cfg(all(feature = "all", target_os = "linux"))))]
     pub fn set_ip_transparent(&self, transparent: bool) -> io::Result<()> {
         unsafe {
@@ -1737,12 +1737,9 @@ impl Socket {
     ///
     /// This returns the value of `TCP_KEEPALIVE` on macOS and iOS and `TCP_KEEPIDLE` on all other
     /// supported Unix operating systems.
-    #[cfg(any(
-        doc,
-        all(
-            feature = "all",
-            not(any(windows, target_os = "haiku", target_os = "openbsd"))
-        )
+    #[cfg(all(
+        feature = "all",
+        not(any(windows, target_os = "haiku", target_os = "openbsd"))
     ))]
     #[cfg_attr(
         docsrs,
@@ -1763,7 +1760,6 @@ impl Socket {
     #[cfg(all(
         feature = "all",
         any(
-            doc,
             target_os = "android",
             target_os = "dragonfly",
             target_os = "freebsd",
@@ -1805,7 +1801,6 @@ impl Socket {
     #[cfg(all(
         feature = "all",
         any(
-            doc,
             target_os = "android",
             target_os = "dragonfly",
             target_os = "freebsd",

--- a/src/sys/unix.rs
+++ b/src/sys/unix.rs
@@ -609,7 +609,8 @@ pub(crate) fn socket(family: c_int, ty: c_int, protocol: c_int) -> io::Result<So
     syscall!(socket(family, ty, protocol))
 }
 
-#[cfg(feature = "all")]
+#[cfg(all(feature = "all", unix))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "all", unix))))]
 pub(crate) fn socketpair(family: c_int, ty: c_int, protocol: c_int) -> io::Result<[Socket; 2]> {
     let mut fds = [0, 0];
     syscall!(socketpair(family, ty, protocol, fds.as_mut_ptr())).map(|_| fds)
@@ -898,8 +899,11 @@ fn into_timeval(duration: Option<Duration>) -> libc::timeval {
     }
 }
 
-#[cfg(feature = "all")]
-#[cfg(not(any(target_os = "haiku", target_os = "openbsd")))]
+#[cfg(all(feature = "all", not(any(target_os = "haiku", target_os = "openbsd"))))]
+#[cfg_attr(
+    docsrs,
+    doc(cfg(all(feature = "all", not(any(target_os = "haiku", target_os = "openbsd")))))
+)]
 pub(crate) fn keepalive_time(fd: Socket) -> io::Result<Duration> {
     unsafe {
         getsockopt::<c_int>(fd, IPPROTO_TCP, KEEPALIVE_TIME)
@@ -1158,7 +1162,7 @@ impl crate::Socket {
     }
 
     /// Sets `SO_NOSIGPIPE` on the socket.
-    #[cfg(all(feature = "all", any(doc, target_vendor = "apple")))]
+    #[cfg(all(feature = "all", target_vendor = "apple"))]
     #[cfg_attr(docsrs, doc(cfg(all(feature = "all", target_vendor = "apple"))))]
     pub fn set_nosigpipe(&self, nosigpipe: bool) -> io::Result<()> {
         self._set_nosigpipe(nosigpipe)

--- a/tests/socket.rs
+++ b/tests/socket.rs
@@ -1367,7 +1367,7 @@ fn tcp_congestion() {
     let cur_tcp_ca = cur_tcp_ca.splitn(2, |num| *num == 0).next().unwrap();
     const OPTIONS: [&[u8]; 2] = [
         b"cubic",
-        #[cfg(target_os = "linux")] // or Android.
+        #[cfg(target_os = "linux")]
         b"reno",
         #[cfg(target_os = "freebsd")]
         b"newreno",

